projects/batfish-common-protocol/src/main/java/org/batfish/common/plugin/PluginConsumer.java

```diff
@@ -150,153 +124,12 @@ private boolean isJavaSerializationData(PushbackInputStream stream) throws IOExc
-    Class<?> pluginClass = null;
-    try {
-      cl.loadClass(className);
-      pluginClass = Class.forName(className, true, cl);
-    } catch (ClassNotFoundException | NoClassDefFoundError e) {
-      // Ignoring this exception is a potentially dangerous hack
-      // but I don't quite know yet why some classes cannot be loaded and how to do things cleanly
-      getLogger().warn("Couldn't load class " + className + ". Skipping.\n");
-      return;
-    }
-    if (!Plugin.class.isAssignableFrom(pluginClass)
-        || Modifier.isAbstract(pluginClass.getModifiers())) {
-      return;
-    }
-    Constructor<?> pluginConstructor;
-    try {
-      pluginConstructor = pluginClass.getConstructor();
-    } catch (NoSuchMethodException | SecurityException e) {
-      throw new BatfishException(
-          "Could not find default constructor in plugin: '" + className + "'", e);
-    }
-    Object pluginObj;
-    try {
-      pluginObj = pluginConstructor.newInstance();
-    } catch (InstantiationException
-        | IllegalAccessException
-        | IllegalArgumentException
-        | InvocationTargetException e) {
-      throw new BatfishException(
-          "Could not instantiate plugin '" + className + "' from constructor", e);
-    }
-    Plugin plugin = (Plugin) pluginObj;
-    plugin.initialize(this);
-  }
-
-  private class JarVisitor extends SimpleFileVisitor<Path> {
-    boolean _hasJars = false;
-    boolean _hasClasses = false;
-
-    @Override
-    public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
-      if (loadPluginJar(path)) {
-        _hasJars = true;
-      } else if (path.toString().endsWith(CLASS_EXTENSION)) {
-        _hasClasses = true;
-      }
-      return FileVisitResult.CONTINUE;
-    }
-  }
-
   protected final void loadPlugins() {
-    // this supports loading plugins either from a directory of jars, or
-    // from a folder of classfiles (useful for eclipse development)
-    for (Path pluginDir : _pluginDirs) {
-      if (Files.exists(pluginDir)) {
-        JarVisitor jarVisitor = new JarVisitor();
-        try {
-          // first load any jars
-          Files.walkFileTree(pluginDir, jarVisitor);
-        } catch (IOException e) {
-          throw new BatfishException("Error walking through plugin dir: '" + pluginDir + "'", e);
-        }
-
-        // if there are class files and no jars, then try to load as a folder
-        if (jarVisitor._hasClasses && !jarVisitor._hasJars) {
-          try {
-            loadPluginFolder(pluginDir);
-          } catch (IOException e) {
-            throw new BatfishException(
-                "Error loading plugin folder: '" + pluginDir.toString() + "'", e);
-          }
-        }
-      }
+    Set<Plugin> plugins =
+        new LinkedHashSet<>(
+            Lists.newArrayList(ServiceLoader.load(Plugin.class, _currentClassLoader)));
+    for (Plugin plugin : plugins) {
+      plugin.initialize(this);
```

Comment:
I'd probably wrap this in a try/catch and report an error loading any of these but continue loading others?

Reply:
Hmm I don't know. I don't think it really makes sense to keep running when a plugin has failed. How about catch, wrap, and throw?

Reply:
Addressed via catching of ServiceConfigurationException, throwing CompositeBatfishException after all are done loading.